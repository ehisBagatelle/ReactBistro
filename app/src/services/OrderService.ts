import { taxes } from "../../test/fixture";
import { formatDecimal } from "../helper";

export type LineItem = {
  name: string;
  category: string;
  price: number;
};

export type Tax = {
  name: string;
  rate: number;
  appliesToCategories: string[];
};

export type Discount = {
  id: number
  type: "dollar" | "percentage";
  amount: number;
};

export type Bill = {
  totalTaxes: string;
  totalDiscounts: string;
  subTotal: string;
  total: string;
};

export interface IOrderService {
  calculateBill(lineItems: LineItem[], appliedDiscounts: Discount[]): Bill;
}

export class OrderService implements IOrderService {
  taxes: Tax[] = taxes;

  public calculateBill(
    lineItems: LineItem[],
    appliedDiscounts: Discount[]
  ): Bill {
    let subTotal = 0;

    // Calculate pre-tax total
    lineItems.forEach((item) => {
      subTotal += item.price;
    });

    // Calculate total taxes
    const totalTaxes = this.calculateTotalTaxes(lineItems);
    
    // Calculate post-tax total
    const postTaxTotal = subTotal + totalTaxes;

    // Calculate total discounts
    const totalDiscounts = this.calculateTotalDiscounts(
      appliedDiscounts,
      postTaxTotal
    );

    const total = postTaxTotal - totalDiscounts

    return {
      totalTaxes: formatDecimal(totalTaxes),
      totalDiscounts: formatDecimal(totalDiscounts),
      subTotal: formatDecimal(subTotal),
      total: formatDecimal(total)
    };
  }

  private calculateTotalTaxes(lineItems: LineItem[]): number {
    let totalTaxes = 0;

    lineItems.forEach((item) => {
      const applicableTaxes = this.taxes.filter((tax) =>
        tax.appliesToCategories.includes(item.category)
      );

      applicableTaxes.forEach((tax) => {
        totalTaxes += item.price * tax.rate;
      });
    });

    return totalTaxes;
  }

  private calculateTotalDiscounts(
    appliedDiscounts: Discount[],
    postTaxTotal: number
  ): number {
    // The order of applied discounts matter and can change the final output totals
    // We recommend going dollar amount discounts first
    let postDiscountTotal = postTaxTotal;
  
    // Separate dollar amount discounts and percentage discounts
    const dollarAmountDiscounts = appliedDiscounts.filter(
      (discount) => discount.type === "dollar"
    );
  
    const percentageDiscounts = appliedDiscounts.filter(
      (discount) => discount.type === "percentage"
    );
  
    // Apply dollar amount discounts first
    dollarAmountDiscounts.forEach((discount) => {
        postDiscountTotal -= discount.amount;
    });
  
    // Apply percentage discounts after dollar amount discounts
    percentageDiscounts.forEach((discount) => {
        postDiscountTotal -= (discount.amount / 100) * postDiscountTotal;
    });
  
    return postTaxTotal - postDiscountTotal;
  }
}
