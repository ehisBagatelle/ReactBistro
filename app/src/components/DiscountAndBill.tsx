import React from "react";
import { View, TouchableOpacity, Text } from "react-native";
import { styles } from "./Components.styles";
import { Title } from "./Title";
import { LineItem } from "../services/OrderService";

interface Props {
  onPressDiscount: () => void;
  order: LineItem[];
}

const DiscountAndBill: React.FC<Props> = ({ onPressDiscount, order }) => (
  <View style={styles.discountAndBill}>
    <View style={styles.discount}>
      <TouchableOpacity onPress={onPressDiscount}>
        <Title
          name="Tap To Add Discount"
          iconSource={require("../../assets/icons8-discount-50.png")}
        />
      </TouchableOpacity>
    </View>
    <View style={styles.bill}>
      <View style={styles.billItem}>
        <Text style={styles.billItemText}>Subtotal</Text>
        <Text style={styles.billItemAmount}>$12.99</Text>
      </View>
      <View style={styles.billItem}>
        <Text style={styles.billItemText}>Discounts</Text>
        <Text style={styles.billItemAmount}>$12.99</Text>
      </View>
      <View style={styles.billItem}>
        <Text style={styles.billItemText}>Tax</Text>
        <Text style={styles.billItemAmount}>$12.99</Text>
      </View>
      <View style={styles.billItem}>
        <Text
          style={styles.billItemText}
        >{`Total (${order.length} item)`}</Text>
        <Text style={styles.billItemAmount}>$12.99</Text>
      </View>
    </View>
  </View>
);

export default DiscountAndBill;
