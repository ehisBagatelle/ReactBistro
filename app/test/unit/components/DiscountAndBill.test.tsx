import React from 'react';
import renderer from 'react-test-renderer';
import DiscountAndBill from '../../../src/components/DiscountAndBill';
import { Bill, LineItem } from '../../../src/services/OrderService';

const mockOnPressDiscount = jest.fn();

const sampleItem: LineItem = {
  name: 'Test Item',
  category: 'Test Category',
  price: 9.99,
  key: '003'
};

const sampleBill: Bill = {
  subTotal: "29.97",
  totalDiscounts: "5.00",
  totalTaxes: "2.00",
  total: "27.97",
};

describe('DiscountAndBill', () => {
  it('renders correctly with discount and bill data', () => {
    const tree = renderer.create(
      <DiscountAndBill onPressDiscount={mockOnPressDiscount} order={[sampleItem, sampleItem]} bill={sampleBill} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
