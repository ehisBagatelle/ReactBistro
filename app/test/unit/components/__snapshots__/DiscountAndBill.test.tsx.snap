// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiscountAndBill renders correctly with discount and bill data 1`] = `
<View
  style={
    {
      "alignSelf": "flex-end",
      "width": "55%",
    }
  }
>
  <View
    style={{}}
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#F5F5F5",
            "display": "flex",
            "flexDirection": "row",
            "height": 45,
            "justifyContent": "space-between",
            "paddingHorizontal": 10,
            "width": "100%",
          }
        }
      >
        <Text>
          Tap To Add Discount
        </Text>
        <Image
          source={
            {
              "testUri": "../../../assets/icons8-discount-50.png",
            }
          }
        />
      </View>
    </View>
  </View>
  <View
    style={
      {
        "padding": 15,
        "width": "100%",
      }
    }
  >
    <View
      style={
        {
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <Text
        style={
          {
            "fontSize": 22,
            "fontStyle": "normal",
            "fontWeight": "400",
          }
        }
      >
        Subtotal
      </Text>
      <Text
        style={
          {
            "fontSize": 22,
            "fontStyle": "normal",
            "fontWeight": "700",
          }
        }
      >
        $
        29.97
      </Text>
    </View>
    <View
      style={
        {
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <Text
        style={
          {
            "fontSize": 22,
            "fontStyle": "normal",
            "fontWeight": "400",
          }
        }
      >
        Discounts
      </Text>
      <Text
        style={
          {
            "fontSize": 22,
            "fontStyle": "normal",
            "fontWeight": "700",
          }
        }
      >
        $
        5.00
      </Text>
    </View>
    <View
      style={
        {
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <Text
        style={
          {
            "fontSize": 22,
            "fontStyle": "normal",
            "fontWeight": "400",
          }
        }
      >
        Tax
      </Text>
      <Text
        style={
          {
            "fontSize": 22,
            "fontStyle": "normal",
            "fontWeight": "700",
          }
        }
      >
        $
        2.00
      </Text>
    </View>
    <View
      style={
        {
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <Text
        style={
          {
            "fontSize": 22,
            "fontStyle": "normal",
            "fontWeight": "400",
          }
        }
      >
        Total (2 item)
      </Text>
      <Text
        style={
          {
            "fontSize": 22,
            "fontStyle": "normal",
            "fontWeight": "700",
          }
        }
      >
        $
        27.97
      </Text>
    </View>
  </View>
</View>
`;
